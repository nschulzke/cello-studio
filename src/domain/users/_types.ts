export * from './sessions/_types';
export * from './profiles/_types';
