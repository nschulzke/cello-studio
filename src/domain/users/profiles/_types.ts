export enum ContactType {
  NONE = 0,
  CALL = 1,
  EMAIL = 2,
  TEXT = 4,
  GROUP_TEXT = 8,
}

export interface Profile {
  name: string;
  parentName: string;
  contactEmail: string;
  contactPhone: string;
  contactType: ContactType;
}
