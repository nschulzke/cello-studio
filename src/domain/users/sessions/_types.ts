export enum Permissions {
  ADMIN = 'ADMIN',
  STUDENT = 'STUDENT',
  NONE = 'NONE',
}

export interface CredentialsRaw {
  email: string;
  password: string;
}

export interface CredentialsHashed {
  email: string;
  hash: string;
}
