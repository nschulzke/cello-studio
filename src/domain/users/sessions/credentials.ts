import * as bcrypt from 'bcrypt';

export const credentials = {
  hash: ({ email, password }) => ({
    email,
    hash: bcrypt.hashSync(password, 10)
  }),

  compare: ({ email: sentEmail, password }, { email: trueEmail, hash } = { email: false, hash: '' }) =>
    sentEmail === trueEmail
    && bcrypt.compareSync(password, hash),
}
