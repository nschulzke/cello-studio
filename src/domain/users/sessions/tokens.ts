import * as jwt from 'jsonwebtoken';

export const prepareTokens = ({ secret, duration }) => ({
  encode: (payload) =>
    jwt.sign({ data: payload }, secret, { expiresIn: duration }),

  decode: (token) => {
    try {
      return jwt.verify(token, secret).data;
    } catch (e) {
      return false;
    }
  },
})
