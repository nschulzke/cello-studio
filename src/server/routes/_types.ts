export * from './api/_types';
