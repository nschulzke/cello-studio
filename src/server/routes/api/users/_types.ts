import { Permissions } from 'domain/users/sessions/_types';
import { Profile } from 'domain/types';
import { TypeChecker } from 'server/helpers/types';

type LoginRequest = {
  email: string;
  password: string;
};

const LoginRequest: TypeChecker<LoginRequest> = {
  is: (obj: any): obj is LoginRequest => {
    return typeof obj.email === 'string'
      && typeof obj.password === 'string';
  },
}

type RegisterRequest = {
  email: string;
  password: string;
};

const RegisterRequest: TypeChecker<RegisterRequest> = {
  is: (obj: any): obj is RegisterRequest => {
    return typeof obj.email === 'string'
      && typeof obj.password === 'string';
  },
}

type UpdateProfileRequest = { email: string, profile: object };
const UpdateProfileRequest: TypeChecker<UpdateProfileRequest> = {
  is: (obj: any): obj is UpdateProfileRequest => {
    return typeof obj.email === 'string'
      && typeof obj.profile === 'object';
  }
}

type LoginResponse = {
  email: string,
  token: string,
  permissions: Permissions,
  profile: Profile,
};

type UpdateProfileResponse = {
  profile: Profile
};

export {
  LoginRequest,
  RegisterRequest,
  UpdateProfileRequest,
  LoginResponse,
  UpdateProfileResponse
};
