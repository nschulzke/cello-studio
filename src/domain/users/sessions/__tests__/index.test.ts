import { prepareUsers } from '../index';
import { prepareTokens } from '../tokens';
import { credentials } from '../credentials';
import { Permissions } from '../_types';

describe('users', () => {
  let newUser = { email: 'test@example.com', password: 'mynewpass' };
  let userModel = { [newUser.email]: { name: 'Test' } };
  let fullUserModel = { [newUser.email]: { ...credentials.hash(newUser), name: 'Test', permissions: Permissions.STUDENT } };
  let tokens = prepareTokens({ secret: 'mysecret', duration: '1h' });
  let users = prepareUsers({ credentials, tokens });

  describe('register', () => {
    it('registers a new user successfully', () => {
      let result = users.register(userModel, newUser);
      expect(result.success).toBe(true);
      let newUserModel = result.data;
      expect(newUserModel[newUser.email].email).toEqual(newUser.email);
      expect(newUserModel[newUser.email].permissions).toEqual(Permissions.STUDENT);
    });

    it('does not modify any existing properties', () => {
      let result = users.register(userModel, newUser);
      expect(result.success).toBe(true);
      let newUserModel = result.data;
      expect(newUserModel[newUser.email].name).toEqual(userModel[newUser.email].name);
    })

    it('rejects if the user is already registered', () => {
      let failure = users.register(fullUserModel, newUser);
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });

    it('rejects an invalid email address', () => {
      let failure = users.register(fullUserModel, { email: 'email', password: 'testing' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('login', () => {
    it('accepts correct login information', () => {
      let result = users.login(fullUserModel, newUser)
      expect(result.success).toBe(true);
      expect(typeof result.data).toBe('object');
      expect(typeof result.data.token).toBe('string');
      expect(result.data.email).toEqual(newUser.email);
      expect(result.data.permissions).toEqual(Permissions.STUDENT);
    });

    it('rejects a bad email address', () => {
      let failure = users.login(fullUserModel, { ...newUser, email: 'hi!' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });

    it('rejects a bad email password', () => {
      let failure = users.login(fullUserModel, { ...newUser, password: 'hi!' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('fetchUser', () => {
    let token = tokens.encode(newUser.email);

    it('returns the user object', () => {
      expect(users.fetchUser(fullUserModel, token).data).toEqual(fullUserModel[newUser.email]);
    });

    it('rejects invalid tokens', () => {
      let failure = users.fetchUser(fullUserModel, 'badtoken');
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('permissions', () => {
    it('returns the permissions for a user', () => {
      let token = tokens.encode(newUser.email);
      expect(users.permissions(fullUserModel, token)).toEqual(Permissions.STUDENT);
    });

    it('returns none for an invalid user', () => {
      let token = tokens.encode('garbage');
      expect(users.permissions(fullUserModel, token)).toEqual(Permissions.NONE);
    });

    it('returns none for an invalid token', () => {
      let token = 'garbage';
      expect(users.permissions(fullUserModel, token)).toEqual(Permissions.NONE);
    });
  });
});
