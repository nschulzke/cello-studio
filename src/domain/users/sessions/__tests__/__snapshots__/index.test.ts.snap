// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users login rejects a bad email address 1`] = `"Invalid email or password."`;

exports[`users login rejects a bad email password 1`] = `"Invalid email or password."`;

exports[`users register rejects an invalid email address 1`] = `"Email is invalid."`;

exports[`users register rejects if the user is already registered 1`] = `"Email is in use."`;

exports[`users verify rejects invalid tokens 1`] = `"Your token could not be validated."`;
